win32/tests/threadpool_win32_int/threadpool_win32_int.c

```diff
@@ -1058,6 +1191,69 @@ TEST_FUNCTION(chaos_knight_test)
     execution_engine_dec_ref(execution_engine);
 }
 
+// test used for detect race condition between timer_restart/timer_cancel and timer destory, failed due to the race condition for the current code, will uncomment after the fix
+//TEST_FUNCTION(chaos_knight_test_with_timers_no_lock)
+//{
+//    // start a number of threads and each of them will do a random action on the threadpool
+//    EXECUTION_ENGINE_PARAMETERS execution_engine_parameters = { 4, 0 };
+//    EXECUTION_ENGINE_HANDLE execution_engine = execution_engine_create(&execution_engine_parameters);
+//    HANDLE thread_handles[CHAOS_THREAD_COUNT];
+//    size_t i;
+//    CHAOS_TEST_DATA chaos_test_data;
+//
+//    THANDLE(THREADPOOL) threadpool = threadpool_create(execution_engine);
+//    ASSERT_IS_NOT_NULL(threadpool);
+//
+//    THANDLE_INITIALIZE_MOVE(THREADPOOL)(&chaos_test_data.threadpool, &threadpool);
+//
+//    (void)InterlockedExchange64(&chaos_test_data.expected_call_count, 0);
+//    (void)InterlockedExchange64(&chaos_test_data.executed_work_functions, 0);
+//    (void)InterlockedExchange64(&chaos_test_data.executed_timer_functions, 0);
+//    (void)InterlockedExchange(&chaos_test_data.timers_starting, 0);
+//    (void)InterlockedExchange(&chaos_test_data.chaos_test_done, 0);
+//    (void)InterlockedExchange(&chaos_test_data.can_start_timers, 1);
+//
+//    for (i = 0; i < MAX_TIMER_COUNT; i++)
+//    {
+//        chaos_test_data.timers[i].timer = NULL;
+//        InterlockedExchange(&chaos_test_data.timers[i].state, TIMER_STATE_NONE);
+//    }
+//
+//    chaos_test_data.work_item_context = threadpool_create_work_item(chaos_test_data.threadpool, work_function, (void*)&chaos_test_data.executed_work_functions);
```

Comment:
>work_item_context [](http://example.com/codeflow?start=22&length=17)

There should be an assert on this not being NULL.

Reply:
this one is NULL in the chaos test because this test only test if it would crash, schedule work item with non-NULL context was tested in the following several test functions.

Reply:
I am confused, doesn't threadpool_create_work_item return something non-NULL always?

Reply:
changed to return a non-NULL context and added the assert.

Reply:
threadpool_create_work_item returns NULL when threadpool is not open, the existing test didn't open the threadpool before this line when I copied from the old test, so get a NULL work_item and didn't actually schedule any work item in the chaos test, the old test directly goes into break in the case of schedule work item/schedule work due to the NULL check. I changed it to return a non-NULL context in the create_work_item and actually run some work items now in the chaos test.